spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/bigquant/Quant.scala

```diff
@@ -0,0 +1,180 @@
+    require(size.length == 2, s"only support 2-dim matrix")
+    require(size.product == (end - start), s"number of elements does not match")
+
+    val height = size(0)
+    val width = size(1)
+
+    val max = new Array[Float](height)
+    val min = new Array[Float](height)
+
+    for (i <- 0 until height) {
+      val maxAndMin = quantize(src, start + i * width, start + (i + 1) * width, dst,
+        dstOffset + i * width)
+
+      max(i) = maxAndMin._1
+      min(i) = maxAndMin._2
+    }
+
+    (max, min)
+  }
+
+  def dequantize(data: Array[Float], start: Int, end: Int, quantizedData: Array[Byte], offset: Int,
+    max: Array[Float], min: Array[Float], size: Array[Int]): Unit = {
+    require(max.length == min.length, s"the number of max doesn't match with the number of min")
+    require(size.length == 2, s"only support 2-dim matrix")
+    require(max.length == size(0),
+      s"the number of max(${max.length}) doesn't match the size(${size(1)})")
+
+    require(size.product == (end - start), s"number of elements does not match")
+
+    val height = size(0)
+    val width = size(1)
+
+    for (i <- 0 until height) {
+      dequantize(data, start + i * width, start + (i + 1) * width,
+        quantizedData, offset + i * width, max(i), min(i))
+    }
+  }
+
+  private[bigdl] def get2Dim(shape: Array[Int]): Array[Int] = {
+    require(shape.length > 1, s"error size dimension, which must be great than 1")
+    val first = shape(0)
+    val last = shape.slice(1, shape.length).product
+    Array(first, last)
+  }
+
+  def quantize(input: Tensor[Float], buffer: Array[Byte],
+    offset: Int): (Array[Float], Array[Float]) = {
+    val length = input.nElement()
+
+    input.dim() match {
+      case 1 =>
+        val (max, min) = quantize(input.storage().array(), input.storageOffset() - 1,
+          length, buffer, offset)
+        (Array(max), Array(min))
+      case x if x > 1 =>
+        val size = get2Dim(input.size())
+        val start = input.storageOffset() - 1
+        val end = start + length
+        val (max, min) = quantize(input.storage().array(), start, end, buffer, offset, size)
+        (max, min)
+      case _ => throw new UnsupportedOperationException(s"unsupported input")
+    }
+  }
+
+  def dequantize(input: Tensor[Float], buffer: Array[Byte], offset: Int, max: Array[Float],
+    min: Array[Float]): Unit = {
+    val start = input.storageOffset() - 1
+    val end = start + input.nElement()
+
+    input.dim() match {
+      case 1 => dequantize(input.storage().array(), start, end, buffer,
+        offset, max(0), min(0))
+      case x if x > 1 =>
+        dequantize(input.storage().array(), start, end, buffer,
+          offset, max, min, get2Dim(input.size()))
+      case _ => throw new UnsupportedOperationException {
+        s"unsupported input dim ${input.dim()}"
+      }
+    }
+  }
+
+  def loss(before: Array[Float], after: Array[Float], start: Int, end: Int): Double = {
```

Comment:
why do we need loss?

Reply:
This is a testing method, I'll make it to private.